model/src/com/redhat/ceylon/model/typechecker/model/Type.java

```diff
@@ -217,11 +217,137 @@ else if (isTypeParameter()) {
      * given type? 
      */
     public boolean isExactly(Type type) {
+        
+        //shortcircuit certain simple cases
+        //to avoid having to resolve aliases
+        //and canonicalize, which can result
+        //in an overflow
+        if (type.isClassOrInterface()) {
+            if (isTypeParameter()) {
+                return false;
+            }
+            if (type.isExactlyNothing()) {
+                return isExactlyNothing();
+            }
+            if (isUnion()) {
+                boolean result = false;
+                for (Type ct: getCaseTypes()) {
+                    if (ct.isExactly(type)) {
+                        result = true;
+                    }
+                    else if (!ct.isSubtypeOf(type)) {
+                        return false;
+                    }
+                }
+                return result;
+            }
+            if (isIntersection()) {
+                boolean result = false;
+                for (Type st: getSatisfiedTypes()) {
+                    if (st.isExactly(type)) {
+                        result = true;
+                    }
+                    else if (!st.isSupertypeOf(type)) {
+                        return false;
+                    }
+                }
+                if (isNotGeneric(type.getDeclaration())) {
+                    return result;
```

Comment:
I can't seem to exercise this code, but could the test be weakened to

```java
    if (result || isNotGeneric(type.getDeclaration())) { return result; }
```

IOW, always let a `true` result be sufficient. I assume this is for stuff like:

    Cov<A> & Cov<B> == Cov<A&B> // equal but requires full algo
    Cov<A> & Cov<B> & Cov<A&B> == Cov<A&B> // ok
    Inv<String> & Inv<Integer> == Nothing // full algo

I could easily be missing something here...